### PR TITLE
Correctly detect the project runtime based on the active language server

### LINKED ISSUE #75: Use correct JDK runtime for projects
Maven builds  fail due to mvn by default using the javahome env variable. However, different projects can be configured to use different runtimes. Furthermore, i myself have 5 jdks installed, my default my javahome points to jdk20. However, my projects do not use java 20. They use a range of versions from 11 to 17, which are explicitly configured in jdtls runtimes. Where jdtls is made aware of which runtime it must use. (or the user can provide all available runtimes on the system, and jdtls can deduce the default one to use, when it scans the gradle or pom files)

Before starting the mvn jobstart, one must export the correct javahome variable pointing to the projects default jdk. That can be obtained by either querying jdtls or as a fallback scanning the pom.

The runtime locations available on the system could be user configureable in the plugin config in case the fallback route is taken. 

The primary idea, however, is to go with a jdtls query, which will return the currently configured default runtime and whrere on the system it is located at. That will then be used to set the javahome in jobstart / termopen.

### PR DESCRIPTION
**Pull request by asmodeus812**
The goal of this pr is to correctly resolve the runtime home directory of the jdk that a specific project is using, after that the JAVA_HOME env variable is set to that, and the appropriate maven or gradle commands can be run. 

This has to be done when we call maven or gradle. The reason this is needed is pretty simple, noone has just one single runtime installed on their system, and most people work with multiple java projects configured/developed to work with different runtime versions. 

For example say a user has installed the latest jdk & setup his JAVA_HOME and PATH to point to jdk21, however the user would like to run/compile the tests for an older java 8, 11 or 17 spring project. 

The JAVA_HOME variable has to be set to the `correct runtime for the project before` any relevant maven or gradle commands are run otherwise, for more simple projects you might get by and compile it, but for more complex projects which might be using deprecated or straight up removed features between major jdk versions, you will either fail to compile the code or you will encounter a runtime exception while running the code.

By default maven always picks up JAVA_HOME, you can also configure this in the pom itself, (by adding additional toolchain and compiler plugins) by setting the path to the runtime in the compiler plugin (for your local machine) gradle is more versatile it still reads JAVA_HOME, however you can start it with `-D` argument or you can configure the path to the runtime in build.gradle. 

Both maven and gradle have toolchain plugins which allow them to specify the java version/runtime for the project separate from the java stack on user's PATH, `manually, proivde absolute path to the desired runtime on your local machine, that the project has to be run with`,  however in practice noone does this, locally, since most projects are microservices, they rely on pod's env.

You can see why this is kind of needed, while it is possible to provide the path to the runtime for a project, and have it be separate from JAVA_HOME (i.e the java stack which is on your PATH)
- would not be deployeable/portable in a cluster
- it is not usually done, since it requires user input, locally modifying the pom
- it is specific for the user's machine, people install them in different places

The solution in this plugin is to provide the runtimes just once, either as env variables JAVA_HOME_XX, or in your lua config, not having to modify the pom for each project, locally, separately, having to add additional maven/gradle plugins to the project just to be able to specify this etc, and you would certainly forget to do that.

TL;DR - Nothing guarantees you that the java stack on user's PATH is the one a specific project requires to be compiled / depends on.

### CODE REVIEW

File: lua/neotest-java/lsp/init.lua [hunk 1]
```diff
@@ -0,0 +1,76 @@
+local log = require("neotest-java.logger")
+local nio = require("nio")
+
+-- table that holds the language server settings, this is mostly done for interfacing with coc.nvim, since native neovim lsp clients hold their settings in the clients table
+local SETTINGS = {}
+
+local function execute_command(command, bufnr)
```
Comment (rcasia) [inline lua/neotest-java/lsp/init.lua:7]: I would separate this two conditional branches into two different functions as their logic is unrelated and they are long blocks of code.

File: lua/neotest-java/lsp/init.lua [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+		if not ok or not result then
+			log.warn(
+				string.format(
+					"Unable to run lsp %s command with payload %s",
+					command.command,
+					vim.inspect(command.arguments)
+				)
+			)
+		end
+		local error = not ok and result ~= vim.NIL and { message = result } or nil
```
Comment (rcasia) [inline lua/neotest-java/lsp/init.lua:35]: Better to replace it by `err` or something else.

**Review by rcasia — CHANGES_REQUESTED** [review]

**Comment by asmodeus812** [comment]
@rcasia could you take a look again, there are a few standing questions, regarding piping, extracting the bufnr and what do we do with gradle's runtime

File: lua/neotest-java/command/runtime.lua
Comment (rcasia) [inline lua/neotest-java/command/runtime.lua:1]: I would not do this job here as the lsp should know what java executable to use using `vscode.java.resolveJavaExecutable` like in here: https://github.com/mfussenegger/nvim-jdtls/blob/6bfd1591583b02e742fc3a2f43393c4ea3b6d3c7/lua/jdtls/util.lua#L70

File: README.md [hunk 3]
```diff
@@ -66,13 +66,21 @@ require("neotest").setup({
   adapters = {
     require("neotest-java")({
         ignore_wrapper = false, -- whether to ignore maven/gradle wrapper
+        java_runtimes = {
```
Comment (rcasia) [inline README.md:69]: At least nvim-jdtls and coc-java already do this work. This seems like duplicate work as lsp could be already has this setting. Or am I missing something?

File: lua/neotest-java/util/read_xml_tag.lua [hunk 4]
```diff
@@ -18,9 +17,11 @@ local read_xml_tag = memo(function(filepath, selector)
 		parsed = parsed[tag]
 	end
 
-	if type(parsed) == "table" then
-		return nil
-	end
+	-- why ? there are tags which can hold multiple tags / array of tags ?
```
Comment (rcasia) [inline lua/neotest-java/util/read_xml_tag.lua:20]: I just wanted to retturn a non-table value, but feel free to use it to retrieve tables aswell.

**Review by rcasia — CHANGES_REQUESTED** [review]

**Comment by rcasia** [comment]
With those changes it looks alright and ready to merge.

**Comment by asmodeus812** [comment]
> With those changes it looks alright and ready to merge.

Not yet, almost. There are a few things to iron out.

**Comment by rcasia** [comment]
I will close this PR as #153 will fix the issue.

Thanks, this PR actually helped to get close to a solution.

**Comment by asmodeus812** [comment]
@rcasia what about porting this to support coc, as the original pr ?

**Comment by rcasia** [comment]
Responded in https://github.com/rcasia/neotest-java/issues/75#issuecomment-2397574152
